src/components/PlaybackControls.tsx

```diff
@@ -0,0 +1,249 @@
+    displayTimes;
+  return `${
+    Math.max(
+      roundedFirstFrameTime.toString().length,
+      roundedLastFrameTime.toString().length,
+      roundedTimeStep.toString().length
+    ) + 1.5
+  }ch`;
+};
+
+const PlayBackControls = (props: PlayBackProps): JSX.Element => {
+  const { controller, playbackData, playbackState, setPlaybackState } = props;
+  const [timeInput, setTimeInput] = useState<number>(0);
+
+  const { timeUnits, firstFrameTime, lastFrameTime, timeStep } = playbackData;
+  const { isPlaying, currentTime } = playbackState;
+
+  const nextFrameUnavailable = currentTime + timeStep >= lastFrameTime;
+  const previousFrameUnavailable = firstFrameTime + timeStep > currentTime;
+
+  const displayTimes = getDisplayTimes(playbackState, playbackData);
+  const timeInputWidth = getTimeInputWidth(displayTimes);
+
+  // useeffect to reset playback state when the last frame is reached
+  useEffect(() => {
+    controller.pause();
+    controller.gotoTime(firstFrameTime);
+    setPlaybackState({
+      currentTime: firstFrameTime,
+      isPlaying: false,
+    });
+  }, [nextFrameUnavailable]);
+
+  const skipToTime = (time: number) => {
+    const isTimePastLastFrame =
+      compareTimes(time, lastFrameTime, timeStep) === 1;
+    const isTimeBeforeFirstFrame =
+      compareTimes(time, firstFrameTime, timeStep) === -1;
+    if (isTimePastLastFrame || isTimeBeforeFirstFrame) {
+      return;
+    }
+    controller.gotoTime(time);
+  };
+
+  const playHandler = (timeOverride?: number) => {
+    const timeOverrideProvided = timeOverride !== undefined;
+    const newTime = timeOverrideProvided ? timeOverride : currentTime;
+    // this is the same as nextFrameUnavailable, but computing it here
+    // prevents a looping bug where playback gets stuck oscillating between
+    // first and last frame
+    const nextFrameIsPastLastFrame = newTime + timeStep >= lastFrameTime;
+    controller.pause();
+    if (nextFrameIsPastLastFrame) {
+      setPlaybackState({
+        isPlaying: false,
+        currentTime: firstFrameTime,
+      });
+      controller.gotoTime(firstFrameTime);
+    } else {
+      setPlaybackState({
+        isPlaying: true,
+        currentTime: newTime,
+      });
+      controller.playFromTime(newTime);
+      controller.resume();
+    }
+  };
+
+  const pauseHandler = () => {
+    controller.pause();
+    setPlaybackState({ ...playbackState, isPlaying: false });
+  };
+
+  const gotoNextFrame = () => {
+    controller.gotoTime(currentTime + timeStep);
+  };
+
+  const goToPreviousFrame = () => {
+    controller.gotoTime(currentTime - timeStep);
+  };
+
+  const handleTimeInputChange = (userInput: number | null): void => {
+    if (userInput !== null) {
+      setTimeInput(userInput as number);
+    }
+  };
+
+  const handleTimeInputKeyDown = (
+    event: React.KeyboardEvent<HTMLInputElement>
+  ): void => {
+    if (event.key === 'Enter' || event.key === 'Tab') {
```

Comment:
If you're checking for tab navigation or the user clicking away, you could also add a handler for the `onBlur` event